prune/prune.go

```diff
@@ -16,184 +16,204 @@ package prune
-type Config struct {
-	Clientset      kubernetes.Interface      // kube client used by pruning
-	LabelSelector  string                    //selector resources to prune
-	DryRun         bool                      //true only performs a check, not removals
-	Resources      []schema.GroupVersionKind //pods, jobs are supported
-	Namespaces     []string                  //empty means all namespaces
-	Strategy       StrategyConfig            //strategy for pruning, either age or max
-	CustomStrategy StrategyFunc              //custom strategy
-	PreDeleteHook  PreDelete                 //called before resource is deleteds
-	Log            logr.Logger               //optional: to overwrite the logger set at context level
+// ErrUnprunable indicates that it is not allowed to prune a specific object.
+type ErrUnprunable struct {
+	Obj    *client.Object
+	Reason string
 }
 
-// Execute causes the pruning work to be executed based on its configuration
-func (config Config) Execute(ctx context.Context) error {
-	log := Logger(ctx, config)
-	log.V(1).Info("Execute Prune")
+// Error returns a string representation of an `ErrUnprunable` error.
+func (e *ErrUnprunable) Error() string {
+	return fmt.Sprintf("%s -- %+v", e.Reason, e.Obj)
+}
+
+// StrategyFunc takes a list of resources and returns the subset to prune.
+type StrategyFunc func(ctx context.Context, objs []client.Object) ([]client.Object, error)
+
+// IsPrunableFunc is a function that checks the data of an object to see whether or not it is safe to prune it.
+// It should return `nil` if it is safe to prune, `ErrUnprunable` if it is unsafe, or another error.
+// It should safely assert the object is the expected type, otherwise it might panic.
+type IsPrunableFunc func(obj client.Object) error
 
-	err := config.validate()
+// PrunerOption configures the pruner.
+type PrunerOption func(p *Pruner)
+
+// NewPruner returns a pruner that uses the given strategy to prune objects.
+func NewPruner(config *rest.Config, opts ...PrunerOption) (Pruner, error) {
+	prunerClient, err := client.New(config, client.Options{})
 	if err != nil {
-		return err
+		return Pruner{}, err
 	}
 
-	for i := 0; i < len(config.Resources); i++ {
-		var resourceList []ResourceInfo
-		var err error
+	pruner := Pruner{
+		Registry: &defaultRegistry,
+		Client:   prunerClient,
+		DryRun:   false,
+		Logger:   Logger(context.Background(), Pruner{}),
+	}
 
-		if config.Resources[i].Kind == PodKind {
-			resourceList, err = config.getSucceededPods(ctx)
-			if err != nil {
-				return err
-			}
-			log.V(1).Info("pods ", "count", len(resourceList))
-		} else if config.Resources[i].Kind == JobKind {
-			resourceList, err = config.getCompletedJobs(ctx)
-			if err != nil {
-				return err
-			}
-			log.V(1).Info("jobs ", "count", len(resourceList))
-		}
+	// Populate the default IsPrunableFunc(s)
+	RegisterIsPrunableFunc(schema.GroupVersionKind{
+		Group:   "core",
+		Version: "v1",
+		Kind:    "Pod",
+	}, DefaultPodIsPrunable)
```

Comment:
These GVKs should be defined somewhere in the kubernetes API. I can help you look if you can't find them, but I don't remember where exactly they are off the top of my head.

Reply:
Replaced this (and all other manual GVK creation of GVKs that should already exist) with:
```go
podGVK := corev1.SchemeGroupVersion.WithKind("Pod")
```
Let me know what you think